File: Java/src/RecipeBuilder.java
package net.glitchtechs.starapi;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.material.MaterialData;
import org.bukkit.plugin.java.JavaPlugin;

public class RecipeBuilder {
    private final JavaPlugin plugin;
    private final ShapedRecipe recipe;
    private final Map<Character, RecipeBuilder.RecipeIngredient> ingredients;

    public RecipeBuilder(JavaPlugin plugin, ItemStack result, String key) {
        NamespacedKey recipeKey = new NamespacedKey(plugin, key);
        this.recipe = new ShapedRecipe(recipeKey, result);
        this.ingredients = new HashMap();
        this.plugin = plugin;
    }

    public RecipeBuilder setShape(String... shape) {
        this.recipe.shape(shape);
        return this;
    }

    public RecipeBuilder setIngredient(char key, Material material, int amount) {
        RecipeBuilder.RecipeIngredient recipeIngredient = new RecipeBuilder.RecipeIngredient(material, amount);
        this.recipe.setIngredient(key, recipeIngredient.getMaterialData());
        this.ingredients.put(key, recipeIngredient);
        return this;
    }

    public RecipeBuilder setIngredient(char key, ItemStack itemStack, int amount) {
        RecipeBuilder.RecipeIngredient recipeIngredient = new RecipeBuilder.RecipeIngredient(itemStack, amount);
        this.recipe.setIngredient(key, recipeIngredient.getMaterialData());
        this.ingredients.put(key, recipeIngredient);
        return this;
    }

    public ShapedRecipe build() {
        Iterator var1 = this.ingredients.entrySet().iterator();

        while(var1.hasNext()) {
            Entry<Character, RecipeBuilder.RecipeIngredient> entry = (Entry)var1.next();
            this.recipe.setIngredient((Character)entry.getKey(), ((RecipeBuilder.RecipeIngredient)entry.getValue()).getMaterialData());
        }

        return this.recipe;
    }

    private static class RecipeIngredient {
        private final ItemStack itemStack;

        public RecipeIngredient(Material material, int amount) {
            this.itemStack = new ItemStack(material, amount);
        }

        public RecipeIngredient(ItemStack itemStack, int amount) {
            this.itemStack = itemStack.clone();
            this.itemStack.setAmount(amount);
        }

        public MaterialData getMaterialData() {
            return this.itemStack.getData();
        }

        public int getAmount() {
            return this.itemStack.getAmount();
        }

        public boolean isSameItem(ItemStack other) {
            return this.itemStack.isSimilar(other);
        }
    }
}

File: Kotlin/src/RecipeBuilder.kt
package net.glitchtechs.starapi

import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.material.MaterialData
import org.bukkit.plugin.java.JavaPlugin


class RecipeBuilder(plugin: JavaPlugin, result: ItemStack?, key: String?) {
    private val plugin: JavaPlugin
    private val recipe: ShapedRecipe
    private val ingredients: MutableMap<Char?, RecipeIngredient?>

    init {
        val recipeKey = NamespacedKey(plugin, key!!)
        this.recipe = ShapedRecipe(recipeKey, result!!)
        this.ingredients = HashMap()
        this.plugin = plugin
    }

    fun setShape(vararg shape: String?): RecipeBuilder {
        recipe.shape(*shape)
        return this
    }

    fun setIngredient(key: Char, material: Material?, amount: Int): RecipeBuilder {
        val recipeIngredient = RecipeIngredient(material, amount)
        recipe.setIngredient(key, recipeIngredient.materialData!!)
        ingredients[key] = recipeIngredient
        return this
    }

    fun setIngredient(key: Char, itemStack: ItemStack, amount: Int): RecipeBuilder {
        val recipeIngredient = RecipeIngredient(itemStack, amount)
        recipe.setIngredient(key, recipeIngredient.materialData!!)
        ingredients[key] = recipeIngredient
        return this
    }

    fun build(): ShapedRecipe {
        val var1: Iterator<*> = ingredients.entries.iterator()

        while (var1.hasNext()) {
            val entry: Map.Entry<Char, RecipeIngredient> = var1.next() as Map.Entry<Char, RecipeIngredient>
            recipe.setIngredient(entry.key, entry.value.materialData!!)
        }

        return this.recipe
    }

    private class RecipeIngredient {
        private val itemStack: ItemStack

        constructor(material: Material?, amount: Int) {
            this.itemStack = ItemStack(material!!, amount)
        }

        constructor(itemStack: ItemStack, amount: Int) {
            this.itemStack = itemStack.clone()
            this.itemStack.amount = amount
        }

        val materialData: MaterialData?
            get() = itemStack.data

        val amount: Int
            get() = itemStack.amount

        fun isSameItem(other: ItemStack?): Boolean {
            return itemStack.isSimilar(other)
        }
    }
}